addons/snailmail/static/src/messaging/component/message/message_tests.js

```diff
@@ -0,0 +1,578 @@
+        id: 10,
+        message_type: 'snailmail',
+        notifications: [['insert', [{
+            failure_type: 'sn_format',
+            id: 11,
+            notification_status: 'exception',
+            notification_type: 'snail',
+        }]]],
+    });
+    await this.createMessageComponent(message);
+
+    assert.containsOnce(
+        document.body,
+        '.o_Message',
+        "should display a message component"
+    );
+    assert.containsOnce(
+        document.body,
+        '.o_Message_notificationIcon',
+        "should display the notification icon"
+    );
+    assert.containsOnce(
+        document.body,
+        '.o_Message_notificationIcon .fa-paper-plane',
+        "icon should represent snailmail"
+    );
+
+    await afterNextRender(() => {
+        document.querySelector('.o_Message_notificationIcon').click();
+    });
+    assert.verifySteps(['do_action'], "'do_action' should have been called");
+});
+
+QUnit.test('Missing Required Fields', async function (assert) {
+    assert.expect(8);
+
+    await this.start({
+        intercepts: {
+            do_action(ev) {
+                assert.step('do_action');
+                assert.strictEqual(
+                    ev.data.action,
+                    'snailmail.snailmail_letter_missing_required_fields_action',
+                    "action should be the one for missing fields"
+                );
+                assert.strictEqual(
+                    ev.data.options.additional_context.letter_id,
+                    22,
+                    "action should have correct letter id"
+                );
+            },
+        },
+        async mockRPC(route, args) {
+            if (args.model === 'snailmail.letter' && args.method === 'search') {
+                assert.step('search');
+                return [22];
+            }
+            return this._super(...arguments);
+        },
+    });
+    const message = this.env.entities.Message.create({
+        id: 10,
+        message_type: 'snailmail',
+        notifications: [['insert', [{
+            failure_type: 'sn_fields',
+            id: 11,
+            notification_status: 'exception',
+            notification_type: 'snail',
+        }]]],
+    });
+    await this.createMessageComponent(message);
+
+    assert.containsOnce(
+        document.body,
+        '.o_Message',
+        "should display a message component"
+    );
+    assert.containsOnce(
+        document.body,
+        '.o_Message_notificationIcon',
+        "should display the notification icon"
+    );
+    assert.containsOnce(
+        document.body,
+        '.o_Message_notificationIcon .fa-paper-plane',
+        "icon should represent snailmail"
+    );
+
+    await afterNextRender(() => {
+        document.querySelector('.o_Message_notificationIcon').click();
+    });
+    assert.verifySteps(
+        ['search', 'do_action'],
+        "'search' and 'do_action' should have been called"
+    );
```

Comment:
at a glance most tests looks a bit too technical in my opinion

Reply:
Well I guess this is due to the result of do_action not being testable, can't really do anything about that here.